### ISSUE TITLE
[How to] Access a JS callback method from the native code using an Asynchronous JS FFI function.

### ISSUE THREAD

**Issue by ganeshrvel** [Koromix/rygel#14]
I have an app which is written in Electron.js 18, I use Go code for the native code. I use CGO to generate the headers for the FFI. I have a requirement where I need to Asynchronous JS functions on the app while the native code (Go lang) needs to access the callback method (and return values) passed down to it via koffi. 

JS koffi implementation
```js
    const FfiPreprocesCb = koffi.callback("FfiPreprocesCb", "void", ["string"]);
    const UploadFiles = lib.func("UploadFiles", "void", [
      "string",
      koffi.pointer(FfiPreprocesCb),
    ]);

    let ret = UploadFiles.async(
      "{}",
      (str) => {
        console.log("output of FfiPreprocesCb", str);
      },
      (err, res) => {
        console.log("Result:", res);
      }
    );
```

Native code (Go/CGO)
```go
 	typedef void (* cb_result_t)(char*);
	void send_result(int64_t ptr, char* json) {
		cb_result_t cb = (cb_result_t) ptr;

		if(cb != 0 && cb != NULL){
			cb(json);
		}
	}
```

In the code above I am passing down the pointer of `FfiPreprocesCb` to the native code which then sends back information to js at regular interval of time. Everything works fine until i use the async method

`let ret = UploadFiles(...)` -> works fine but is blocking code

`let ret = UploadFiles.async(` -> DOES NOT work.

It throws an error:
```
#
# Fatal error in v8::HandleScope::CreateHandle()
# Cannot create a handle without a HandleScope
#
```

I have a few questions here:
1. How do I make an Asynchronous JS FFI call and also pass a pointer to a JS callback function which can bring back information from the native code?
2. When I make a Synchronous FFI call, it throws an error: `error Command failed with signal "SIGSEGV".`. What needs to be done here to end the session gracefully?
3. The native code runs for long and send back information periodically, how do I make sure that the pointer to the JS callback function doesnt get auto cleaned?
4. Do I need to manually free the allocated pointer?

Thanks!

**Comment by nwrkbiz** [comment]
Hi,

can you try the following (with the new koffi 2.2.0 released today):

```js
// the typedef in koffi syntax
const cb_result_t = koffi.callback('void cb_result_t(char*)');

// registered callback
const rawCB = koffi.register(function(result){
        // convert result char* to string and print to console
        console.log(koffi.decode(result, 'char*'));
    }, koffi.pointer(cb_result_t));

// async function
const send_result = lib.func('void send_result(cb_result_t* ptr, char* json');

// ...
// ...
// ... Anywhere in your code:
send_result(rawCB, "{}");
```

**Comment by ganeshrvel** [comment]
Hi, I got it working but unfortunately there is just one more hiccup, the FFI code is blocking the main thread.


```js
    setInterval(() => {
      console.log("inside the setInterval")
    }, 1000)

    const cb_result_t = koffi.callback('void cb_result_t(char*)');
    const rawCB = koffi.register(function (result) {
      // console.log(result);
      // console.log(koffi.decode(result, "char*"));
    }, koffi.pointer(cb_result_t));

    const send_result = this.lib.func('void UploadFiles(char* json, cb_result_t* ptr)');

    send_result("{}", rawCB );

    setTimeout(() => {
      console.log("inside the settimeout")
    }, 5000)
```

How do we make this asynchronous and make sure that  the main thread isnt blocked?

**Comment by ganeshrvel** [comment]
@nwrkbiz 
I tried to use `send_result.async()` to make the above code work asynchronously but it threw a `HandleScope` error.

Any help here will be highly appreciated.

**Comment by nwrkbiz** [comment]
Can you provide more of your CGO code? Are you using threads in CGO?

**Comment by ganeshrvel** [comment]
Ah I see, I wasnt using threads in Go, I thought that koffi function will be async despite of how the native program is written.

I have added a go thread (routine) and yes, now the node.js script's main thread isnt getting blocked but now the app is crashing with the error: `error Command failed with signal "SIGSEGV".

Looks like a thread safety issue. I am debugging it further to see what's going on here.

CGO code:
```go
/*
	typedef void (* on_cb_result_t)(char*);
	void send_cb_result(on_cb_result_t* ptr, char* json) {
		on_cb_result_t cb = (on_cb_result_t) ptr;

		if(cb != 0 && cb != NULL){
			cb(json);
		}
	}
*/
import "C"

//export UploadFiles
func UploadFiles(json *C.char, cbs *C.on_cb_result_t) 
	go func() {
		for {
			C.send_cb_result(cbs, C.CString("some string"))

			time.Sleep(time.Millisecond * 500)
		}
	}()
}
``` 

Generated header-file.h:
```h
   typedef void (* on_cb_result_t)(char*);
   extern void UploadFiles(char* json, on_cb_result_t* cbs);
```

Javascript code:
```js
    setInterval(() => {
      console.log("inside the setInterval");
    }, 1000);

    const cb_result_t = koffi.callback("void on_cb_result_t(char*)");
    const rawCB = koffi.register((result) => {
      console.log(result);
      // console.log(koffi.decode(result, "char*"));
    }, koffi.pointer(cb_result_t));

    const send_result = this.lib.func(
      "void UploadFiles(char* json, on_cb_result_t* cbs)"
    );

    send_result("{}", rawCB);

    setTimeout(() => {
      console.log("inside the settimeout");
    }, 5000);

    koffi.unregister(rawCB);

```

**Comment by nwrkbiz** [comment]
Please see my example within the last post, I think that covers what you want to archieve.

**Comment by ganeshrvel** [comment]
@nwrkbiz Thanks for the code. 

As you see in this comment: https://github.com/Koromix/rygel/issues/14#issuecomment-1362980189 my CPU intrinsic task is happening in the Go side and I am just using a C wrapper to pass the progress information back to js. Adding Go routine has made the ffi code work asynchronously but a `SIGSEGV` error is being thrown: `error Command failed with signal "SIGSEGV"`. Looks like a thread safety issue.

I am debugging it further.

**Comment by Koromix** [comment]
I'll look into this over the week-end if all goes well :)

**Comment by ganeshrvel** [comment]
@Koromix @nwrkbiz 

I have some interesting find:

FFI code:
```js
   setInterval(() => {
      console.log("inside the JS setInterval");
    }, 1000);

    const TestAsync = this.lib.func('void TestAsync()');

    TestAsync.async((err, res) => {
      console.log('Result:', res);
    })

    return Promise.resolve();
```

generated header.h
```c
extern void TestAsync();
```

Go code:
```go
//export TestAsync
func TestAsync() {
	for {
		time.Sleep(time.Millisecond * 500)

		log.Printf("Go TestAsync blocking code")
	}
}
```

Output:
```shell
2022/12/24 13:19:18 Go TestAsync blocking code 
inside the JS setInterval
2022/12/24 13:19:18 Go TestAsync blocking code 
2022/12/24 13:19:19 Go TestAsync blocking code 
inside the JS setInterval
2022/12/24 13:19:19 Go TestAsync blocking code 
2022/12/24 13:19:20 Go TestAsync blocking code 
inside the JS setInterval
2022/12/24 13:19:20 Go TestAsync blocking code 
2022/12/24 13:19:21 Go TestAsync blocking code 
inside the JS setInterval
2022/12/24 13:19:21 Go TestAsync blocking code 
2022/12/24 13:19:22 Go TestAsync blocking code 
inside the settimeout
inside the JS setInterval
2022/12/24 13:19:22 Go TestAsync blocking code 
2022/12/24 13:19:23 Go TestAsync blocking code 
inside the JS setInterval
2022/12/24 13:19:23 Go TestAsync blocking code 
2022/12/24 13:19:24 Go TestAsync blocking code
```

If you notice the above output, the `async` KOFFI function allows the execution of the continuous and blocking Native (GO) code asynchronously. No threads were used here.

While the koffi `register` and `callback` functions does NOT allow asynchronous execution of the blocking Native (GO) code when no threads are used.

But when I use the Go routine (threads), the koffi `register` and `callback` functions work asynchronously but a `SIGSEGV` error is being thrown: `error Command failed with signal "SIGSEGV"`. **Looks like a thread safety issue.**

A code sample:
```js
    setInterval(() => {
      console.log("inside the setInterval")
    }, 1000)

    const cb_result_t = koffi.callback('void on_cb_result_t(char*)');
    const rawCB = koffi.register(function (result) {
      // console.log(result);
      // console.log(koffi.decode(result, "char*"));
    }, koffi.pointer(cb_result_t));

    const send_result = this.lib.func('void UploadFiles(char* json, cb_result_t* ptr)');

    send_result("{}", rawCB );

    setTimeout(() => {
      console.log("inside the settimeout")
    }, 5000)
```

Generated header-file.h:
```h
   typedef void (* on_cb_result_t)(char*);
   extern void UploadFiles(char* json, on_cb_result_t* cbs);
```


CGO code:
```go
/*
	typedef void (* on_cb_result_t)(char*);
	void send_cb_result(on_cb_result_t* ptr, char* json) {
		on_cb_result_t cb = (on_cb_result_t) ptr;

		if(cb != 0 && cb != NULL){
			cb(json);
		}
	}
*/
import "C"

//export UploadFiles
func UploadFiles(json *C.char, cbs *C.on_cb_result_t) 
    for {
	    C.send_cb_result(cbs, C.CString("some string"))
	    time.Sleep(time.Millisecond * 500)
    }
}
```

**Comment by ganeshrvel** [comment]
@Koromix any help here will be highly appreciated

**Comment by ganeshrvel** [comment]
@Koromix Not trying to pressurize you here, just checking if there are any updates on this? Thank you.

**Comment by Koromix** [comment]
Node.js is inherently single-thread, hence why it does not work or crash V8.

What I need to add is support for async callbacks, which will be called on the main thread (even though the Go code runs on another thread) even through the native async function runs on another thread.

This will work but please note that the callback will block until the main JS thread is free to process it (= you await something else, run the event loop, etc.).

I plan to work on this in the next few days, now that the holidays are over.

**Comment by ganeshrvel** [comment]
@Koromix thanks for the updates. Looking forward for it. I am looking for a way to move out of ffi-napi. It has some very bad js GC issue. koffi is the only viable alternative that I see in my case.

**Comment by Koromix** [comment]
I've pushed Koffi 2.2.2-beta.2, install with `npm install koffi@beta`.

This version should automatically run callbacks on the main thread when inside an async function, so the JS code you posted in your first post should work with Koffi 2.2.2.

```js
const FfiPreprocesCb = koffi.callback("FfiPreprocesCb", "void", ["string"]);
const UploadFiles = lib.func("UploadFiles", "void", [
  "string",
  koffi.pointer(FfiPreprocesCb),
]);

let ret = UploadFiles.async(
  "{}",
  (str) => {
    console.log("output of FfiPreprocesCb", str);
  },
  (err, res) => {
    console.log("Result:", res);
  }
);
```

Please let me know how that goes ;)

**Comment by ganeshrvel** [comment]
@Koromix 
Thank you, the above async callback code worked. But when I reference the callback function pointer (of `on_cb_result_t`) from a `Goroutine` (threads) a SIGSEGV error is being thrown: `error Command failed with signal "SIGSEGV"`. Looks like when I use threads inside the native code and try to access the js koffi callback function, a thread safety issue occurs.


This issue never happened while using the ffi-napi package. This only happens which I use threads in the native code and try to call the callback function pointer. 


Sample Go Code with threads:
```Go
/*
	typedef void (* on_cb_result_t)(char*);
	void send_cb_result(on_cb_result_t* ptr, char* json) {
		on_cb_result_t cb = (on_cb_result_t) ptr;

		if(cb != 0 && cb != NULL){
			cb(json);
		}
	}
*/
import "C"

//export UploadFiles
func UploadFiles(json *C.char, cbs *C.on_cb_result_t) 
	go func() {
		for {
			C.send_cb_result(cbs, C.CString("some string"))

			time.Sleep(time.Millisecond * 500)
		}
	}()
}
```

**Comment by Koromix** [comment]
Oh yes, I see.

Koffi uses static trampoline, which have some advantages (they work on more platforms, especially hardened ones) but here it's a problem, because I cannot pass any context along. 

Koffi uses thread-local variables to work around this...
... but it obviously fails in this case. 

I have some ideas to work around this, I'll keep you informed today ;)

**Comment by ganeshrvel** [comment]
Oh man @Koromix, You are simply awesome! ;) 
If you could get this working then I will replace ffi-napi with koffi in one of my FOSS apps [OpenMTP](https://github.com/ganeshrvel/openmtp).

**Comment by Koromix** [comment]
I've pushed 2.2.2-beta.3, it should work 🤞

Or at least the [reproduction test](https://github.com/Koromix/rygel/blob/b5816ec75f2ba69bdf324fb5e4824315714010ac/src/koffi/test/callbacks.js#L265-L273) works, unlike in 2.2.2-beta.2 where it crashed the same way your code did.

Let me know how it goes :)

**Comment by ganeshrvel** [comment]
The code threw this error: `[Error: Cannot use non-registered callback beyond FFI call]`

**Comment by Koromix** [comment]
I've pushed 2.2.2-beta.4 to fix the issue with the "Cannot use non-registered callback beyond FFI call" check when a non-registered callback is used from another thread.

Of course, if you want to use a callback beyond the FFI call, you still need to register it. If you haven't read about the two callback types, go here: https://koffi.dev/callbacks#callback-types

**Comment by ganeshrvel** [comment]
@Koromix thank you for the clarification.


So essentially this code is good enough after including the ` koffi.unregister(rawCB)` right?

```
const cb_result_t = koffi.callback('void on_cb_result_t(char*)');
const rawCB = koffi.register(function (result) {
  console.log("Callback result: ", result);
}, koffi.pointer(cb_result_t));

const TestFilesCb = lib.func('void TestFiles(char* json, on_cb_result_t* ptr)');

TestFilesCb.async("{}", rawCB, (err, res) => {
  if (err) {
    console.error("TestFiles err:", err);

    return
  }

  console.log("TestFiles Result:", res);


  koffi.unregister(rawCB); <------ unregister
});
```

**Comment by Koromix** [comment]
It is correct!

But I think you don't need to register in this case, because the native code won't use the callback after the call finishes. You still need 2.2.2-beta.4 to avoid the issue you saw before with non-registered callbacks.

Registered callbacks are needed in other cases. An example would be GUI handlers, where the JS callbacks will be called sometime in the future when events happen (e.g. after the `SetClickEventHandler()` has returned, when the user actually clicks on something).

In your code, the callback gets called while TestFilesCb runs (even if it does in another thread) so it should be alright without registration. But I need to think about it and check the changes I've made a bit more, there might be other issues.

Once I'm happy with the code I'll release 2.2.2.

**Comment by Koromix** [comment]
So this should work with beta.4

```js
const cb_result_t = koffi.callback('void on_cb_result_t(char*)');
const rawCB = function (result) {
  console.log("Callback result: ", result);
};

const TestFilesCb = lib.func('void TestFiles(char* json, on_cb_result_t* ptr)');

TestFilesCb.async("{}", rawCB, (err, res) => {
  if (err) {
    console.error("TestFiles err:", err);

    return
  }

  console.log("TestFiles Result:", res);
});
```

**Comment by ganeshrvel** [comment]
Ah, make sense. Anyway I have decided to go ahead with the registered callback. Thank you for all the efforts.


Interestingly my electron app is throwing a `dyld: lazy symbol binding failed: Symbol not found: _napi_type_tag_object`. I am currently investigating the issue. It worked fine in a normal node.js project.

I will raise a new issue for this.


error:
```
dyld: lazy symbol binding failed: Symbol not found: _napi_type_tag_object
  Referenced from: /path/to/openmtp/node_modules/koffi/src/koffi/build/koffi.node
  Expected in: flat namespace

dyld: Symbol not found: _napi_type_tag_object
  Referenced from: /path/to/openmtp/node_modules/koffi/src/koffi/build/koffi.node
  Expected in: flat namespace

/path/to/openmtp/node_modules/electron/dist/Electron.app/Contents/MacOS/Electron exited with signal SIGABRT
```

**Comment by ganeshrvel** [comment]
Things look fine. I have raised a pr to fix the bundler issue: https://github.com/Koromix/rygel/pull/16

**Comment by Koromix** [comment]
Did you fix the Electron issue? From the error, guess it was because Electron was too "old", Koffi requires N-API version 8.

**Comment by ganeshrvel** [comment]
@Koromix: Oh yes, you are right. I accidentally switched to an older version of electron. It is working fine now. 😃 


Hopefully I can release my app soon to the public with full koffi implementation 👍 .

**Comment by Koromix** [comment]
Thanks, that's great :)
Don't hesitate with any new issue :)

I've just added a N-API version check to index.js.

**Comment by Koromix** [comment]
You should update to 2.2.2-beta.6 which fixes a race condition with the new code, which could lead to a crash or weird failures when calling a callback passed to a secondary thread (a goroutine in your case).

This version also raises the limit from 16 registered callbacks (and 16 temporary callbacks at the same time) to a shared global limit of 1024.

**Comment by Koromix** [comment]
Also, it things are okay on your side, I will release 2.2.2 proper in a couple days, unless a new issue arises.

**Comment by ganeshrvel** [comment]
Thanks for the heads up. So far things seems to be working fine. I will keep looking out for issues.


Thanks

**Comment by Koromix** [comment]
Released 2.2.2 already, because it should be at least as stable as 2.2.1 so why not :)
If something arises I'll just make a new release.

**Comment by ganeshrvel** [comment]
Awesome!
